From 791a6a1835485a621d54f59f33f7bb9d56874200 Mon Sep 17 00:00:00 2001
From: Emilio Pozuelo Monfort <pochu27@gmail.com>
Date: Wed, 20 Jul 2016 17:09:15 +0200
Subject: [PATCH] Fix division by zero when calculating the refresh rate

---
 gdk/x11/gdkscreen-x11.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

Index: gtk+-3.21.6/gdk/x11/gdkscreen-x11.c
===================================================================
--- gtk+-3.21.6.orig/gdk/x11/gdkscreen-x11.c	2016-09-14 04:20:01.880668212 +0200
+++ gtk+-3.21.6/gdk/x11/gdkscreen-x11.c	2016-09-14 04:20:01.876668150 +0200
@@ -636,7 +636,8 @@
               XRRModeInfo *xmode = &resources->modes[j];
               if (xmode->id == crtc->mode)
                 {
-                  refresh_rate = (1000 * xmode->dotClock) / (xmode->hTotal *xmode->vTotal);
+                  if (xmode->hTotal && xmode->vTotal)
+                    refresh_rate = (1000 * xmode->dotClock) / (xmode->hTotal *xmode->vTotal);
                   break;
                 }
             }
