Index: gtk+2.0-2.10.10/configure.in
===================================================================
--- gtk+2.0-2.10.10.orig/configure.in	2007-03-12 15:07:22.000000000 +0100
+++ gtk+2.0-2.10.10/configure.in	2007-03-13 14:10:00.000000000 +0100
@@ -351,7 +351,8 @@
 if test "$os_win32" != yes; then
     # libtool option to control which symbols are exported
     # right now, symbols starting with _ are not exported
-    LIBTOOL_EXPORT_OPTIONS='-export-symbols-regex "^[[^_]].*"'
+    # Disabled until -export-symbols-regex works (RH patch)
+    LIBTOOL_EXPORT_OPTIONS=
 else
     # We currently use .def files on Windows (for gdk-pixbuf, gdk and gtk)
     LIBTOOL_EXPORT_OPTIONS=
