diff -ruN gtk+-2.2.2.orig/docs/reference/gdk/gdk-docs.sgml gtk+-2.2.2/docs/reference/gdk/gdk-docs.sgml
--- gtk+-2.2.2.orig/docs/reference/gdk/gdk-docs.sgml	2002-11-22 05:19:00.000000000 +0900
+++ gtk+-2.2.2/docs/reference/gdk/gdk-docs.sgml	2003-08-05 02:02:01.000000000 +0900
@@ -1,6 +1,6 @@
 <?xml version="1.0"?>
 <!DOCTYPE book PUBLIC "-//OASIS//DTD DocBook XML V4.1.2//EN" 
-               "http://www.oasis-open.org/docbook/xml/4.1.2/docbookx.dtd" [
+               "/usr/share/sgml/docbook/dtd/xml/4.1.2/docbookx.dtd" [
 <!ENTITY gdk-General SYSTEM "xml/general.xml">
 <!ENTITY gdk-Bitmaps-and-Pixmaps SYSTEM "xml/pixmaps.xml">
 <!ENTITY gdk-Images SYSTEM "xml/images.xml">
diff -ruN gtk+-2.2.2.orig/docs/reference/gdk-pixbuf/gdk-pixbuf.sgml gtk+-2.2.2/docs/reference/gdk-pixbuf/gdk-pixbuf.sgml
--- gtk+-2.2.2.orig/docs/reference/gdk-pixbuf/gdk-pixbuf.sgml	2002-11-22 05:19:00.000000000 +0900
+++ gtk+-2.2.2/docs/reference/gdk-pixbuf/gdk-pixbuf.sgml	2003-08-05 02:01:37.000000000 +0900
@@ -1,6 +1,6 @@
 <?xml version="1.0"?>
 <!DOCTYPE book PUBLIC "-//OASIS//DTD DocBook XML V4.1.2//EN" 
-               "http://www.oasis-open.org/docbook/xml/4.1.2/docbookx.dtd" [
+               "/usr/share/sgml/docbook/dtd/xml/4.1.2/docbookx.dtd" [
 <!ENTITY gdk-pixbuf-gdk-pixbuf SYSTEM "xml/gdk-pixbuf.xml">
 <!ENTITY gdk-pixbuf-refcounting SYSTEM "xml/refcounting.xml">
 <!ENTITY gdk-pixbuf-file-loading SYSTEM "xml/file-loading.xml">
diff -ruN gtk+-2.2.2.orig/docs/reference/gtk/gtk-docs.sgml gtk+-2.2.2/docs/reference/gtk/gtk-docs.sgml
--- gtk+-2.2.2.orig/docs/reference/gtk/gtk-docs.sgml	2003-05-06 03:16:06.000000000 +0900
+++ gtk+-2.2.2/docs/reference/gtk/gtk-docs.sgml	2003-08-05 02:00:43.000000000 +0900
@@ -1,6 +1,6 @@
 <?xml version="1.0"?>
 <!DOCTYPE book PUBLIC "-//OASIS//DTD DocBook XML V4.1.2//EN" 
-               "http://www.oasis-open.org/docbook/xml/4.1.2/docbookx.dtd" [
+               "/usr/share/sgml/docbook/dtd/xml/4.1.2/docbookx.dtd" [
 <!ENTITY GtkAccelLabel SYSTEM "xml/gtkaccellabel.xml">
 <!ENTITY GtkAdjustment SYSTEM "xml/gtkadjustment.xml">
 <!ENTITY GtkAlignment SYSTEM "xml/gtkalignment.xml">
