Index: gtk+2.0-2.11.5/configure.in
===================================================================
--- gtk+2.0-2.11.5.orig/configure.in	2007-07-02 17:04:27.000000000 +0200
+++ gtk+2.0-2.11.5/configure.in	2007-07-03 11:34:03.000000000 +0200
@@ -134,6 +134,7 @@
 dnl Initialize libtool
 AC_PROG_CC
 AM_DISABLE_STATIC
+AM_CONDITIONAL([ENABLE_STATIC], [test "$enable_static" = yes])
 
 dnl 
 dnl Check for a working C++ compiler, but do not bail out, if none is found.
Index: gtk+2.0-2.11.5/modules/input/Makefile.am
===================================================================
--- gtk+2.0-2.11.5.orig/modules/input/Makefile.am	2007-07-02 17:04:10.000000000 +0200
+++ gtk+2.0-2.11.5/modules/input/Makefile.am	2007-07-03 11:34:03.000000000 +0200
@@ -106,8 +106,12 @@
 if CROSS_COMPILING
 RUN_QUERY_IMMODULES_TEST=false
 else
+if ENABLE_STATIC
+RUN_QUERY_IMMODULES_TEST=false
+else
 RUN_QUERY_IMMODULES_TEST=test -z "$(DESTDIR)"
 endif
+endif
 
 # Running this if cross compiling or if DESTDIR is set is going to
 # not work at all, so skip it.
@@ -148,7 +152,18 @@
 	$(IM_IME_MODULE)
 
 gtk.immodules: Makefile.am $(module_LTLIBRARIES)
-	$(top_builddir)/gtk/gtk-query-immodules-2.0 $(module_LTLIBRARIES) > gtk.immodules
+	@if $(RUN_QUERY_IMMODULES_TEST) ; then \
+	  echo "$(top_builddir)/gtk/gtk-query-immodules-2.0 > gtk.immodules" ; \
+	$(top_builddir)/gtk/gtk-query-immodules-2.0 $(module_LTLIBRARIES) > gtk.immodules; \
+	else \
+	  echo "***" ; \
+	  echo "*** Warning: gtk.immodules not built" ; \
+	  echo "***" ; \
+	  echo "*** Generate this file manually on host" ; \
+	  echo "*** system using gtk-query-immodules-2.0" ; \
+	  echo "***" ; \
+	  touch gtk.immodules; \
+        fi
 
 CLEANFILES = gtk.immodules
 
