diff -ruN gtk+-2.2.4.orig/aclocal.m4 gtk+-2.2.4/aclocal.m4
--- gtk+-2.2.4.orig/aclocal.m4	2003-09-05 02:17:07.000000000 +0900
+++ gtk+-2.2.4/aclocal.m4	2003-09-08 00:47:50.000000000 +0900
@@ -3123,7 +3123,7 @@
 # This must be Linux ELF.
 linux*)
   case $host_cpu in
-  alpha*|hppa*|i*86|ia64*|m68*|mips*|powerpc*|sparc*|s390*|sh*|x86_64*)
+  alpha*|arm*|hppa*|i*86|ia64*|m68*|mips*|mipsel|powerpc*|sparc*|s390*|sh*|x86_64*)
     lt_cv_deplibs_check_method=pass_all ;;
   *)
     # glibc up to 2.1.1 does not perform some relocations on ARM
diff -ruN gtk+-2.2.4.orig/configure gtk+-2.2.4/configure
--- gtk+-2.2.4.orig/configure	2003-09-08 00:46:04.000000000 +0900
+++ gtk+-2.2.4/configure	2003-09-08 00:49:01.000000000 +0900
@@ -3625,7 +3625,7 @@
 # This must be Linux ELF.
 linux*)
   case $host_cpu in
-  alpha*|hppa*|i*86|ia64*|m68*|mips*|powerpc*|sparc*|s390*|sh*|x86_64*)
+  alpha*|arm*|hppa*|i*86|ia64*|m68*|mips*|mipsel|powerpc*|sparc*|s390*|sh*|x86_64*)
     lt_cv_deplibs_check_method=pass_all ;;
   *)
     # glibc up to 2.1.1 does not perform some relocations on ARM
