Index: gtk+-2.14.7/configure.in
===================================================================
--- gtk+-2.14.7.orig/configure.in	2009-01-07 17:33:57.000000000 +0100
+++ gtk+-2.14.7/configure.in	2009-02-17 23:06:24.718714576 +0100
@@ -152,6 +152,7 @@ esac
 dnl Initialize libtool
 AC_PROG_CC
 AM_DISABLE_STATIC
+AM_CONDITIONAL([ENABLE_STATIC], [test "$enable_static" = yes])
 
 dnl 
 dnl Check for a working C++ compiler, but do not bail out, if none is found.
Index: gtk+-2.14.7/modules/input/Makefile.am
===================================================================
--- gtk+-2.14.7.orig/modules/input/Makefile.am	2009-01-07 17:33:31.000000000 +0100
+++ gtk+-2.14.7/modules/input/Makefile.am	2009-02-17 23:06:24.731083956 +0100
@@ -178,8 +178,12 @@ dist_imconffile_DATA = im-multipress.con
 if CROSS_COMPILING
 RUN_QUERY_IMMODULES_TEST=false
 else
+if ENABLE_STATIC
+RUN_QUERY_IMMODULES_TEST=false
+else
 RUN_QUERY_IMMODULES_TEST=test -z "$(DESTDIR)"
 endif
+endif
 
 # Running this if cross compiling or if DESTDIR is set is going to
 # not work at all, so skip it.
@@ -240,7 +244,18 @@ noinst_LTLIBRARIES =				\
 included-modules: $(noinst_LTLIBRARIES)
 
 gtk.immodules: Makefile.am $(module_LTLIBRARIES)
-	$(top_builddir)/gtk/gtk-query-immodules-2.0 $(module_LTLIBRARIES) > gtk.immodules
+	@if $(RUN_QUERY_IMMODULES_TEST) ; then \
+	  echo "$(top_builddir)/gtk/gtk-query-immodules-2.0 > gtk.immodules" ; \
+	$(top_builddir)/gtk/gtk-query-immodules-2.0 $(module_LTLIBRARIES) > gtk.immodules; \
+	else \
+	  echo "***" ; \
+	  echo "*** Warning: gtk.immodules not built" ; \
+	  echo "***" ; \
+	  echo "*** Generate this file manually on host" ; \
+	  echo "*** system using gtk-query-immodules-2.0" ; \
+	  echo "***" ; \
+	  touch gtk.immodules; \
+        fi
 
 CLEANFILES = gtk.immodules
 
