diff -Nur gtk+-2.8.13.orig/gtk/gtkicontheme.c gtk+-2.8.13/gtk/gtkicontheme.c
--- gtk+-2.8.13.orig/gtk/gtkicontheme.c	2006-02-13 03:33:31.000000000 +0100
+++ gtk+-2.8.13/gtk/gtkicontheme.c	2006-02-26 00:09:20.000000000 +0100
@@ -421,6 +421,8 @@
 			"gtk-fallback-icon-theme", &fallback_theme, NULL);
 	}
 
+      if (!fallback_theme)
+	fallback_theme = g_strdup ("gnome");
       if (!theme)
 	theme = g_strdup (DEFAULT_THEME_NAME);
 
