Description: don't build perf measurement tools in static builds
Bug: http://bugzilla.gnome.org/show_bug.cgi?id=346559

Index: gtk+3.0-2.90.5/Makefile.am
===================================================================
--- gtk+3.0-2.90.5.orig/Makefile.am	2010-06-30 13:37:16.000000000 +0200
+++ gtk+3.0-2.90.5/Makefile.am	2010-07-30 18:57:46.184082799 +0200
@@ -1,7 +1,12 @@
 ## Makefile.am for GTK+
 include $(top_srcdir)/Makefile.decl
 
-SRC_SUBDIRS = gdk gtk modules demos tests perf
+# don't build perf measurement tools in static builds
+if !ENABLE_STATIC
+MAYBE_PERF = perf
+endif
+
+SRC_SUBDIRS = gdk gtk modules demos tests $(MAYBE_PERF)
 SUBDIRS = po po-properties $(SRC_SUBDIRS) docs m4macros build
 
 ACLOCAL_AMFLAGS = -I m4 ${ACLOCAL_FLAGS}
