Description: don't build perf measurement tools in static builds
Bug: http://bugzilla.gnome.org/show_bug.cgi?id=346559

Index: gtk+-2.14.7/Makefile.am
===================================================================
--- gtk+-2.14.7.orig/Makefile.am	2009-01-07 17:33:57.000000000 +0100
+++ gtk+-2.14.7/Makefile.am	2009-02-17 23:07:14.895082593 +0100
@@ -1,7 +1,12 @@
 ## Makefile.am for GTK+
 include $(top_srcdir)/Makefile.decl
 
-SRC_SUBDIRS = gdk-pixbuf gdk gtk modules demos tests perf contrib
+# don't build perf measurement tools in static builds
+if !ENABLE_STATIC
+MAYBE_PERF = perf
+endif
+
+SRC_SUBDIRS = gdk-pixbuf gdk gtk modules demos tests $(MAYBE_PERF) contrib
 SUBDIRS = po po-properties $(SRC_SUBDIRS) docs m4macros
 
 # require automake 1.4
