Index: gtk+2.0-2.10.10/Makefile.am
===================================================================
--- gtk+2.0-2.10.10.orig/Makefile.am	2007-03-12 15:12:11.000000000 +0100
+++ gtk+2.0-2.10.10/Makefile.am	2007-03-13 14:10:07.000000000 +0100
@@ -1,6 +1,11 @@
 ## Makefile.am for GTK+
 
-SRC_SUBDIRS = gdk-pixbuf gdk gtk modules demos tests perf contrib
+# don't build perf measurement tools in static builds
+if !ENABLE_STATIC
+MAYBE_PERF = perf
+endif
+
+SRC_SUBDIRS = gdk-pixbuf gdk gtk modules demos tests $(MAYBE_PERF) contrib
 SUBDIRS = po po-properties $(SRC_SUBDIRS) docs m4macros
 
 # require automake 1.4
