Description: don't build perf measurement tools in static builds
Bug: http://bugzilla.gnome.org/show_bug.cgi?id=346559

Index: b/Makefile.am
===================================================================
--- a/Makefile.am	2010-04-10 14:52:04.949938994 +0200
+++ b/Makefile.am	2010-04-10 14:52:07.085945966 +0200
@@ -1,7 +1,12 @@
 ## Makefile.am for GTK+
 include $(top_srcdir)/Makefile.decl
 
-SRC_SUBDIRS = gdk-pixbuf gdk gtk modules demos tests perf contrib
+# don't build perf measurement tools in static builds
+if !ENABLE_STATIC
+MAYBE_PERF = perf
+endif
+
+SRC_SUBDIRS = gdk-pixbuf gdk gtk modules demos tests $(MAYBE_PERF) contrib
 SUBDIRS = po po-properties $(SRC_SUBDIRS) docs m4macros build
 
 # require automake 1.4
