Index: gtk+2.0-2.15.5/configure.in
===================================================================
--- gtk+2.0-2.15.5.orig/configure.in	2009-03-03 00:53:16.000000000 -0300
+++ gtk+2.0-2.15.5/configure.in	2009-03-03 00:57:09.000000000 -0300
@@ -1703,21 +1703,23 @@
 
 GDK_PACKAGES="$PANGO_PACKAGES gio-2.0"
 if test "x$gdktarget" = "xx11"; then
-  GDK_PACKAGES="$GDK_PACKAGES $X_PACKAGES cairo-xlib"
+  GDK_PRIVATE_PACKAGES="$X_PACKAGES cairo-xlib"
 fi
-GDK_DEP_LIBS="$GDK_EXTRA_LIBS `$PKG_CONFIG --libs $GDK_PIXBUF_PACKAGES $GDK_PACKAGES` $GDK_PIXBUF_EXTRA_LIBS"
-GDK_DEP_CFLAGS="`$PKG_CONFIG --cflags  gthread-2.0 $GDK_PIXBUF_PACKAGES $GDK_PACKAGES` $GDK_PIXBUF_EXTRA_CFLAGS $GDK_EXTRA_CFLAGS"
+GDK_DEP_LIBS="$GDK_EXTRA_LIBS `$PKG_CONFIG --libs $GDK_PIXBUF_PACKAGES $GDK_PACKAGES $GDK_PRIVATE_PACKAGES` $GDK_PIXBUF_EXTRA_LIBS"
+GDK_DEP_CFLAGS="`$PKG_CONFIG --cflags  gthread-2.0 $GDK_PIXBUF_PACKAGES $GDK_PACKAGES $GDK_PRIVATE_PACKAGES` $GDK_PIXBUF_EXTRA_CFLAGS $GDK_EXTRA_CFLAGS"
 
 #
 # If we aren't writing explicit dependencies, then don't put the extra libraries we need
 # into the pkg-config files
 #
 if test $enable_explicit_deps != yes ; then
-  GDK_PACKAGES="$PANGO_PACKAGES"
   GDK_EXTRA_LIBS=
+else
+  GDK_PACKAGES="$GDK_PACKAGES $GDK_PRIVATE_PACKAGES"
 fi
 
 AC_SUBST(GDK_PACKAGES)
+AC_SUBST(GDK_PRIVATE_PACKAGES)
 AC_SUBST(GDK_EXTRA_LIBS)
 AC_SUBST(GDK_EXTRA_CFLAGS)
 AC_SUBST(GDK_DEP_LIBS)
Index: gtk+2.0-2.15.5/gdk-2.0.pc.in
===================================================================
--- gtk+2.0-2.15.5.orig/gdk-2.0.pc.in	2009-03-02 16:03:47.000000000 -0300
+++ gtk+2.0-2.15.5/gdk-2.0.pc.in	2009-03-03 00:53:35.000000000 -0300
@@ -8,5 +8,6 @@
 Description: GTK+ Drawing Kit (${target} target)
 Version: @VERSION@
 Requires: gdk-pixbuf-2.0 @GDK_PACKAGES@
+Requires.private: @GDK_PRIVATE_PACKAGES@
 Libs: -L${libdir} -lgdk-${target}-@GTK_API_VERSION@ @GDK_EXTRA_LIBS@
 Cflags: -I${includedir}/gtk-2.0 -I${libdir}/gtk-2.0/include @GDK_EXTRA_CFLAGS@
