Index: gtk+2.0-2.10.6/gtk/gtkicontheme.c
===================================================================
--- gtk+2.0-2.10.6.orig/gtk/gtkicontheme.c	2006-08-15 20:25:35.000000000 +0200
+++ gtk+2.0-2.10.6/gtk/gtkicontheme.c	2006-12-28 23:41:22.000000000 +0100
@@ -393,6 +393,8 @@
 			"gtk-fallback-icon-theme", &fallback_theme, NULL);
 	}
 
+      if (!fallback_theme)
+	fallback_theme = g_strdup ("gnome");
       if (!theme)
 	theme = g_strdup (DEFAULT_THEME_NAME);
 
