Index: gtk/updateiconcache.c
===================================================================
RCS file: /cvs/gnome/gtk+/gtk/updateiconcache.c,v
retrieving revision 1.4
diff -u -r1.4 updateiconcache.c
--- gtk+/gtk/updateiconcache.c	1 Feb 2005 13:55:07 -0000	1.4
+++ gtk+/gtk/updateiconcache.c	22 Feb 2005 12:43:34 -0000
@@ -17,6 +17,8 @@
  * Boston, MA 02111-1307, USA.
  */
 
+#include <config.h>
+
 #include <stdlib.h>
 #include <stdio.h>
 #include <string.h>
