diff -urN gtk+-2.10.1.orig/gtk/gtkicontheme.c gtk+-2.10.1/gtk/gtkicontheme.c
--- gtk+-2.10.1.orig/gtk/gtkicontheme.c	2006-07-20 20:07:21.000000000 +0200
+++ gtk+-2.10.1/gtk/gtkicontheme.c	2006-08-15 00:06:20.000000000 +0200
@@ -392,6 +392,8 @@
 			"gtk-fallback-icon-theme", &fallback_theme, NULL);
 	}
 
+      if (!fallback_theme)
+	fallback_theme = g_strdup ("gnome");
       if (!theme)
 	theme = g_strdup (DEFAULT_THEME_NAME);
 
