diff --git a/configure.ac b/configure.ac
index 8944f6e..4ab86e4 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1260,10 +1260,11 @@ fi
 CFLAGS="$saved_cflags"
 LDFLAGS="$saved_ldflags"
 
-GDK_PACKAGES="$PANGO_PACKAGES $GIO_PACKAGE $X_PACKAGES $WAYLAND_PACKAGES gdk-pixbuf-2.0 $cairo_backends cairo-gobject"
+GDK_PACKAGES="$PANGO_PACKAGES gdk-pixbuf-2.0"
+GDK_PRIVATE_PACKAGES="$GIO_PACKAGE $X_PACKAGES $WAYLAND_PACKAGES $cairo_backends cairo-gobject"
 
-GDK_DEP_LIBS="$GDK_EXTRA_LIBS `$PKG_CONFIG --libs $GDK_PACKAGES` $MATH_LIB"
-GDK_DEP_CFLAGS="`$PKG_CONFIG --cflags  gthread-2.0 $GDK_PACKAGES` $GDK_EXTRA_CFLAGS"
+GDK_DEP_LIBS="$GDK_EXTRA_LIBS `$PKG_CONFIG --libs $GDK_PACKAGES $GDK_PRIVATE_PACKAGES` $MATH_LIB"
+GDK_DEP_CFLAGS="`$PKG_CONFIG --cflags  gthread-2.0 $GDK_PACKAGES $GDK_PRIVATE_PACKAGES` $GDK_EXTRA_CFLAGS"
 #
 # If we aren't writing explicit dependencies, then don't put the extra libraries we need
 # into the pkg-config files
diff --git a/gdk-3.0.pc.in b/gdk-3.0.pc.in
index 20c54fe..97b0537 100644
--- a/gdk-3.0.pc.in
+++ b/gdk-3.0.pc.in
@@ -8,5 +8,6 @@ Name: GDK
 Description: GTK+ Drawing Kit
 Version: @VERSION@
 Requires: @GDK_PACKAGES@
+Requires.private: @GDK_PRIVATE_PACKAGES@
 Libs: -L${libdir} -lgdk-3 @GDK_EXTRA_LIBS@
 Cflags: -I${includedir}/gtk-@GTK_API_VERSION@ @GDK_EXTRA_CFLAGS@
