From 05abea6c4fb7bf1db0ca72c65b255af975208278 Mon Sep 17 00:00:00 2001
From: Matthias Clasen <mclasen@redhat.com>
Date: Wed, 08 Dec 2010 16:35:06 +0000
Subject: Add missing include

See bug 636732
---
diff --git a/gdk/x11/gdkwindow-x11.h b/gdk/x11/gdkwindow-x11.h
index 0c02b67..85d892b 100644
--- a/gdk/x11/gdkwindow-x11.h
+++ b/gdk/x11/gdkwindow-x11.h
@@ -30,6 +30,8 @@
 #include "gdk/x11/gdkprivate-x11.h"
 #include "gdk/gdkwindowimpl.h"
 
+#include <X11/Xlib.h>
+
 #ifdef HAVE_XDAMAGE
 #include <X11/extensions/Xdamage.h>
 #endif
--
cgit v0.8.3.1
