Index: gtk+2.0-2.11.2/gdk-pixbuf/io-ico.c
===================================================================
--- gtk+2.0-2.11.2.orig/gdk-pixbuf/io-ico.c	2007-06-06 13:43:15.000000000 +0200
+++ gtk+2.0-2.11.2/gdk-pixbuf/io-ico.c	2007-06-13 11:07:11.000000000 +0200
@@ -208,6 +208,14 @@
 	
  	State->HeaderSize = 6 + IconCount*16;
 
+	if (State->HeaderSize < 0) {
+		g_set_error (error,
+			     GDK_PIXBUF_ERROR,
+			     GDK_PIXBUF_ERROR_CORRUPT_IMAGE,
+			     _("Invalid header in icon"));
+		return;
+	}
+
  	if (State->HeaderSize>State->BytesInHeaderBuf) {
  		guchar *tmp=g_try_realloc(State->HeaderBuf,State->HeaderSize);
 		if (!tmp) {
