GNOME #447118; related to Debian #381081

Index: gtk+-2.12.2/configure.in
===================================================================
--- gtk+-2.12.2.orig/configure.in	2007-11-27 23:06:22.533217754 +0100
+++ gtk+-2.12.2/configure.in	2007-11-27 23:06:23.961299135 +0100
@@ -1405,8 +1405,18 @@ if test "x$gdktarget" = "xdirectfb"; the
 
   if $PKG_CONFIG --atleast-version $DIRECTFB_REQUIRED_VERSION directfb && $PKG_CONFIG --exists cairo-directfb ; then
       AC_MSG_RESULT(found)
-      GDK_EXTRA_CFLAGS="`$PKG_CONFIG --cflags cairo-directfb` $GDK_EXTRA_CFLAGS"
-      GDK_EXTRA_LIBS="`$PKG_CONFIG --libs cairo-directfb` $GDK_EXTRA_LIBS"
+      # prepend cairo-directfb so that cairo-directfb's CFLAGS appear before
+      # cairo's if both are available (cairo is required by pango); this is
+      # enough to build gdk/
+      GDK_PACKAGES="cairo-directfb directfb $GDK_PACKAGES"
+      # you'd think this isn't needed, but as soon as cairo-directfb or
+      # directfb requires a "rpath" and you want to build binaries (for
+      # example gtk-query-immodules), you need the CFLAGS and LDFLAGS
+      # everywhere; this trick abuses the fact that GTK_DEP_LIBS and
+      # GTK_DEP_CFLAGS include the GDK_EXTRA_LIBS and _FLAGS to pull the
+      # necessary rpath flags
+      GDK_EXTRA_CFLAGS="`$PKG_CONFIG --cflags cairo-directfb directfb`"
+      GDK_EXTRA_LIBS="`$PKG_CONFIG --libs cairo-directfb directfb` $GDK_EXTRA_LIBS"
   else
       AC_MSG_ERROR([
 *** DirectFB $DIRECTFB_REQUIRED_VERSION or newer and the cairo backend 
@@ -1458,7 +1468,7 @@ fi
 CFLAGS="$saved_cflags"
 LDFLAGS="$saved_ldflags"
 
-GDK_PACKAGES="$PANGO_PACKAGES"
+GDK_PACKAGES="$GDK_PACKAGES $PANGO_PACKAGES"
 if test "x$gdktarget" = "xx11"; then
   GDK_PACKAGES="$GDK_PACKAGES $X_BASE_PACKAGES"
   GDK_PRIVATE_PACKAGES="$X_PACKAGES"
