From 62cbc1acd44ea01e87ece0743db39e184dfa9411 Mon Sep 17 00:00:00 2001
From: Matthias Clasen <mclasen@redhat.com>
Date: Wed, 22 Dec 2010 11:39:58 -0500
Subject: [PATCH] continue to install gdk-$TARGET-3.0.pc

---
 Makefile.am |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index 48ad13a..107985a 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -72,6 +72,7 @@ pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = gdk-3.0.pc gtk+-3.0.pc gail-3.0.pc
 
 pkgconfig_DATA += $(patsubst %,gtk+-%-3.0.pc,@gdktarget@)
+pkgconfig_DATA += $(patsubst %,gdk-%-3.0.pc,@gdktarget@)
 
 if OS_UNIX
 pkgconfig_DATA += gtk+-unix-print-3.0.pc
-- 
1.7.2.3

