Debian #381081; GNOME #447118

Index: gtk+2.0-2.10.10/gdk/directfb/gdkdrawable-directfb.c
===================================================================
--- gtk+2.0-2.10.10.orig/gdk/directfb/gdkdrawable-directfb.c	2007-03-12 15:06:54.000000000 +0100
+++ gtk+2.0-2.10.10/gdk/directfb/gdkdrawable-directfb.c	2007-03-13 14:10:23.000000000 +0100
@@ -45,6 +45,7 @@
 #include "gdkregion-generic.h"
 #include "gdkalias.h"
 
+#include <cairo.h>
 #include "cairo-directfb.h"
 
 #define WARN_UNIMPLEMENTED(func)\
