Index: gtk+2.0-2.10.10/gdk/directfb/gdkgc-directfb.c
===================================================================
--- gtk+2.0-2.10.10.orig/gdk/directfb/gdkgc-directfb.c	2007-03-12 15:06:51.000000000 +0100
+++ gtk+2.0-2.10.10/gdk/directfb/gdkgc-directfb.c	2007-03-13 14:10:26.000000000 +0100
@@ -31,6 +31,7 @@
  */
 
 #include <config.h>
+#undef GDK_DISABLE_DEPRECATED
 #include "gdk.h"
 
 #include <string.h>
