--- gtk+-2.8.20-dfb/gdk/directfb/gdkdrawable-directfb.c.orig	2006-09-10 17:07:48.000000000 +0200
+++ gtk+-2.8.20-dfb/gdk/directfb/gdkdrawable-directfb.c		2006-09-10 17:16:32.000000000 +0200
@@ -428,7 +428,7 @@
           for (i = 0; i < clip->numRects; i++)
             {
               DFBRegion reg = { clip->rects[i].x1,     clip->rects[i].y1,
-                                clip->rects[i].x2, clip->rects[i].y2 };
+                                clip->rects[i].x2 - 1, clip->rects[i].y2 - 1 };
 
               impl->surface->SetClip (impl->surface, &reg);
               impl->surface->TileBlit (impl->surface, surface, NULL, x, y);
@@ -457,7 +457,7 @@
   else
     {
 
-      DFBRegion region = { x, y, x + width+1, y + height+1 };
+      DFBRegion region = { x, y, x + width, y + height };
       impl->surface->SetClip (impl->surface, &region);
 
       /*  DirectFB does not draw rectangles the X way. Using DirectFB,
@@ -518,7 +518,7 @@
             for (i = 0; i < clip->numRects; i++)
               {
                                 DFBRegion reg = { clip->rects[i].x1,     clip->rects[i].y1, 
-                    clip->rects[i].x2 , clip->rects[i].y2  };
+                    clip->rects[i].x2 - 1, clip->rects[i].y2 - 1 };
 
                 impl->surface->SetClip (impl->surface, &reg);
                 impl->surface->FillTriangle (impl->surface,
@@ -624,7 +624,7 @@
   for (i = 0; i < clip->numRects; i++)
     {
       DFBRegion reg = { clip->rects[i].x1,     clip->rects[i].y1,
-                        clip->rects[i].x2 , clip->rects[i].y2 };
+                        clip->rects[i].x2 - 1, clip->rects[i].y2 - 1 };
 
       impl->surface->SetClip (impl->surface, &reg);
       impl->surface->Blit (impl->surface, src_impl->surface, &rect,
@@ -708,7 +708,7 @@
   for (i = 0; i < clip->numRects; i++)
     {
       DFBRegion reg = { clip->rects[i].x1,   clip->rects[i].y1,
-                        clip->rects[i].x2, clip->rects[i].y2 };
+                        clip->rects[i].x2 - 1, clip->rects[i].y2 - 1 };
 
       impl->surface->SetClip (impl->surface, &reg);
 
@@ -819,7 +819,7 @@
   for (i = 0; i < clip->numRects; i++)
     {
       DFBRegion reg = { clip->rects[i].x1,   clip->rects[i].y1,
-                        clip->rects[i].x2, clip->rects[i].y2 };
+                        clip->rects[i].x2 - 1, clip->rects[i].y2 - 1 };
 
       impl->surface->SetClip (impl->surface, &reg);
       impl->surface->DrawLines (impl->surface, lines, npoints - 1);
@@ -871,7 +871,7 @@
       for (i = 0; i < clip->numRects; i++)
         {
           DFBRegion reg = { clip->rects[i].x1,     clip->rects[i].y1,
-                            clip->rects[i].x2 , clip->rects[i].y2  };
+                            clip->rects[i].x2 - 1 , clip->rects[i].y2 - 1 };
 
           impl->surface->SetClip (impl->surface, &reg);
           impl->surface->Blit (impl->surface,
@@ -1037,7 +1037,7 @@
                             GdkRegion               *region)
 {
   DFBRegion reg = { region->extents.x1,     region->extents.y1,
-                    region->extents.x2 , region->extents.y2  };
+                    region->extents.x2 - 1 , region->extents.y2 - 1 };
 
   _gdk_directfb_update (impl, &reg);
 }
