diff -urN gtk+-2.10.1.orig/configure.in gtk+-2.10.1/configure.in
--- gtk+-2.10.1.orig/configure.in	2006-08-15 20:02:19.000000000 +0200
+++ gtk+-2.10.1/configure.in	2006-08-15 20:02:35.000000000 +0200
@@ -337,6 +337,15 @@
 # Honor aclocal flags
 ACLOCAL="$ACLOCAL $ACLOCAL_FLAGS"
 
+# Let the user override the name of the cairo module if he uses another
+# backend; defaults to "cairo" which uses the default cairo backend;
+# cairo-$backend modules Require the cairo module
+AC_ARG_WITH(cairo-backend,
+            AC_HELP_STRING([--with-cairo-backend=BACKEND],
+                           [Select a non-default Cairo backend to use]),
+            cairo_module="cairo-$withval",
+            cairo_module="cairo")
+
 ## Initial sanity check, done here so that users get told they 
 ## have the wrong dependencies as early in the process as possible.
 ## Later on we actually use the cflags/libs from separate pkg-config
@@ -347,7 +356,7 @@
   [glib-2.0 >= glib_required_version dnl
    atk >= atk_required_version dnl
    pango >= pango_required_version dnl
-   cairo >= cairo_required_version])
+   $cairo_module >= cairo_required_version])
 
 if test "$os_win32" != yes; then
     # libtool option to control which symbols are exported
@@ -1521,7 +1530,7 @@
         LIBS="$gtk_save_LIBS"
 fi
 
-GTK_PACKAGES="atk cairo"
+GTK_PACKAGES="atk $cairo_module"
 GTK_EXTRA_LIBS=
 GTK_EXTRA_CFLAGS= 
 GTK_DEP_LIBS="$GDK_EXTRA_LIBS $GTK_DEP_LIBS_FOR_X `$PKG_CONFIG --libs $GDK_PIXBUF_PACKAGES $PANGO_PACKAGES $GTK_PACKAGES_FOR_X $GTK_PACKAGES` $GTK_EXTRA_LIBS $GDK_PIXBUF_EXTRA_LIBS"
