Index: gtk+-2.12.2/configure.in
===================================================================
--- gtk+-2.12.2.orig/configure.in	2007-11-27 23:06:23.961299135 +0100
+++ gtk+-2.12.2/configure.in	2007-11-27 23:06:56.411148347 +0100
@@ -1540,7 +1540,7 @@ PANGO_PREFIX="`$PKG_CONFIG --variable=pr
 CAIRO_PREFIX="`pkg-config --variable=prefix cairo`"
 
 if test $enable_explicit_deps != yes ; then
-  GDK_PIXBUF_EXTRA_LIBS=
+  GDK_PIXBUF_EXTRA_LIBS="$MATH_LIB"
 fi
 
 AC_SUBST(GDK_PIXBUF_PACKAGES)
