Index: b/configure.in
===================================================================
--- a/configure.in	2010-04-10 14:52:08.109945409 +0200
+++ b/configure.in	2010-04-10 14:52:09.177943468 +0200
@@ -1806,7 +1806,7 @@
 CAIRO_PREFIX="`pkg-config --variable=prefix cairo`"
 
 if test $enable_explicit_deps != yes ; then
-  GDK_PIXBUF_EXTRA_LIBS=
+  GDK_PIXBUF_EXTRA_LIBS="$MATH_LIB"
 fi
 
 AC_SUBST(GDK_PIXBUF_PACKAGES)
