diff -urN gtk+-2.10.1-dfb.orig/configure.in gtk+-2.10.1-dfb/configure.in
--- gtk+-2.10.1-dfb.orig/configure.in	2006-08-15 20:20:36.000000000 +0200
+++ gtk+-2.10.1-dfb/configure.in	2006-08-15 21:02:09.000000000 +0200
@@ -1426,8 +1426,8 @@
 
   if $PKG_CONFIG --atleast-version $DIRECTFB_REQUIRED_VERSION directfb ; then
       AC_MSG_RESULT(found)
-      GDK_EXTRA_CFLAGS="`$PKG_CONFIG --cflags directfb` $GDK_EXTRA_CFLAGS"
-      GDK_EXTRA_LIBS="`$PKG_CONFIG --libs directfb` $GDK_EXTRA_LIBS"
+      GDK_EXTRA_CFLAGS="`$PKG_CONFIG --cflags directfb $cairo_module` $GDK_EXTRA_CFLAGS"
+      GDK_EXTRA_LIBS="`$PKG_CONFIG --libs directfb $cairo_module` $GDK_EXTRA_LIBS"
   else
       AC_MSG_ERROR([
 *** DirectFB $DIRECTFB_REQUIRED_VERSION or newer is required. The latest
