Index: gtk+-2.12.1/configure.in
===================================================================
--- gtk+-2.12.1.orig/configure.in	2007-11-18 01:23:24.599043331 +0100
+++ gtk+-2.12.1/configure.in	2007-11-18 01:23:24.695048802 +0100
@@ -1537,7 +1537,7 @@ PANGO_PREFIX="`$PKG_CONFIG --variable=pr
 CAIRO_PREFIX="`pkg-config --variable=prefix cairo`"
 
 if test $enable_explicit_deps != yes ; then
-  GDK_PIXBUF_EXTRA_LIBS=
+  GDK_PIXBUF_EXTRA_LIBS="$MATH_LIB"
 fi
 
 AC_SUBST(GDK_PIXBUF_PACKAGES)
