diff -urN gtk+-2.10.1.orig/gdk/directfb/gdkdrawable-directfb.c gtk+-2.10.1/gdk/directfb/gdkdrawable-directfb.c
--- gtk+-2.10.1.orig/gdk/directfb/gdkdrawable-directfb.c	2006-07-14 02:17:52.000000000 +0200
+++ gtk+-2.10.1/gdk/directfb/gdkdrawable-directfb.c	2006-08-17 14:56:25.000000000 +0200
@@ -64,9 +64,6 @@
                                      GdkPoint    *points,
                                      gint         npoints);
 
-static void inline gdk_directfb_update_region (GdkDrawableImplDirectFB *impl,
-                                               GdkRegion               *region);
-
 static cairo_surface_t *gdk_directfb_ref_cairo_surface (GdkDrawable *drawable);
 
 
@@ -451,7 +448,6 @@
             }
         }
 
-      gdk_directfb_update_region (impl, clip);
       gdk_region_destroy (clip);
     }
   else
@@ -466,8 +462,6 @@
       impl->surface->DrawRectangle (impl->surface,
                                     x, y, width , height);
       impl->surface->SetClip (impl->surface, NULL);
-
-      //_gdk_directfb_update (impl, &region);
     }
 }
 
@@ -528,8 +522,6 @@
 
               }
             impl->surface->SetClip (impl->surface, NULL);
-
-            gdk_directfb_update_region (impl, clip);
             gdk_region_destroy (clip);
 
             return;
@@ -631,7 +623,6 @@
                            xdest, ydest);
     }
   impl->surface->SetClip (impl->surface, NULL);
-  gdk_directfb_update_region (impl, clip);
   gdk_region_destroy (clip);
 }
 
@@ -679,8 +670,6 @@
     }
 
   gdk_region_destroy (clip);
-
-  //_gdk_directfb_update (impl, &region);
 }
 
 static void
@@ -759,8 +748,6 @@
       if (segs->y2 > region.y2)
         region.y2 = segs->y2;
     }
-
-  //_gdk_directfb_update (impl, &region);
 }
 
 static void
@@ -828,8 +815,6 @@
   impl->surface->SetClip (impl->surface, NULL);
 
   gdk_region_destroy (clip);
-
-  //_gdk_directfb_update (impl, &region);
 }
 
 static void
@@ -883,8 +868,6 @@
       image_private->surface->Lock (image_private->surface, DSLF_WRITE,
                                     &image->mem, &pitch);
       image->bpl = pitch;
-
-      gdk_directfb_update_region (impl, clip);
     }
 
   gdk_region_destroy (clip);
@@ -1032,32 +1015,6 @@
   return object_type;
 }
 
-static inline void
-gdk_directfb_update_region (GdkDrawableImplDirectFB *impl,
-                            GdkRegion               *region)
-{
-  DFBRegion reg = { region->extents.x1,     region->extents.y1,
-                    region->extents.x2 , region->extents.y2  };
-
-  _gdk_directfb_update (impl, &reg);
-}
-
-void
-_gdk_directfb_update (GdkDrawableImplDirectFB *impl,
-                      DFBRegion               *region)
-{
-  g_return_if_fail (GDK_IS_DRAWABLE_IMPL_DIRECTFB (impl));
-
-  if (impl->buffered)
-    return;
-
-  if (!impl->surface)
-    return;
-
-  impl->surface->Flip (impl->surface, region, 0);
-}
-
-
 static GdkScreen * gdk_directfb_get_screen (GdkDrawable    *drawable){
         return gdk_screen_get_default();
 }
diff -urN gtk+-2.10.1.orig/gdk/directfb/gdkgeometry-directfb.c gtk+-2.10.1/gdk/directfb/gdkgeometry-directfb.c
--- gtk+-2.10.1.orig/gdk/directfb/gdkgeometry-directfb.c	2006-02-05 05:04:28.000000000 +0100
+++ gtk+-2.10.1/gdk/directfb/gdkgeometry-directfb.c	2006-08-17 14:56:25.000000000 +0200
@@ -119,8 +119,7 @@
               impl->surface->SetClip (impl->surface, &update);
               impl->surface->Blit (impl->surface, impl->surface, NULL, dx, dy);
               impl->surface->SetClip (impl->surface, NULL);
-
-              _gdk_directfb_update (impl, &update);
+              impl->surface->Flip(impl->surface,&update,0);
             }
         }
     }
@@ -238,18 +237,18 @@
 	if (impl->surface)
 	{
     	DFBRectangle source = { dest_extents.x - dx, 
-							 dest_extents.y - dy,
+					    dest_extents.y - dy,
                              dest_extents.width,
                              dest_extents.height};
     	DFBRegion destination = { dest_extents.x, 
-							 	  dest_extents.y,
-                             	  dest_extents.width,
-                             	  dest_extents.height};
+						 dest_extents.y,
+                             	 dest_extents.x+dest_extents.width-1,
+                             	 dest_extents.y+dest_extents.height-1};
 
               impl->surface->SetClip (impl->surface, &destination);
               impl->surface->Blit (impl->surface, impl->surface,&source,dx,dy);
               impl->surface->SetClip (impl->surface, NULL);
-              _gdk_directfb_update (impl, &destination);
+              impl->surface->Flip(impl->surface,&destination,0);
 	}
   gdk_region_destroy (src_region);
   gdk_region_destroy (dest_region);
diff -urN gtk+-2.10.1.orig/gdk/directfb/gdkwindow-directfb.c gtk+-2.10.1/gdk/directfb/gdkwindow-directfb.c
--- gtk+-2.10.1.orig/gdk/directfb/gdkwindow-directfb.c	2006-07-14 02:17:52.000000000 +0200
+++ gtk+-2.10.1/gdk/directfb/gdkwindow-directfb.c	2006-08-17 14:56:25.000000000 +0200
@@ -65,7 +65,6 @@
  */
 static GSList *update_windows = NULL;
 static guint update_idle = 0;
-static gboolean debug_updates = FALSE;
 
 static void
 gdk_window_directfb_process_all_updates (void)
@@ -609,9 +608,6 @@
   if (window == gdk_directfb_focused_window)
     gdk_directfb_change_focus (NULL);
 
-/*
-printf( " DESTROYING WINDOW %p %p %d parent=%p\n",window,impl->window,recursing,(GDK_WINDOW_OBJECT (window))->parent);
-*/
   if (!recursing && !foreign_destroy && impl->window ) {
     	impl->window->SetOpacity (impl->window,0);
    		impl->window->Close(impl->window);
@@ -1298,9 +1294,6 @@
       _gdk_directfb_move_resize_child (list->data,
                                        private->x, private->y,
                                        impl->drawable.width, impl->drawable.height);
-      //FIXEME should this really happen ?
-      if( impl->drawable.surface )
-        impl->drawable.surface->GetPosition(impl->drawable.surface,&x,&y);
     }
 }
 
@@ -2831,8 +2824,6 @@
   
   if (!gdk_region_empty (visible_region))
     {
-      //if (debug_updates)
-       // draw_ugly_color (window, region);
       
       if (private->update_area)
 	{
@@ -2845,7 +2836,6 @@
 	  gdk_window_schedule_update (window);
 	}
     }
-  
     gdk_region_destroy (visible_region);
 }
 
@@ -2880,14 +2870,6 @@
 	  GdkRegion *expose_region;
 	  GdkRegion *window_region;
           gint width, height;
-
-          //if (debug_updates)
-           // {
-              /* Make sure we see the red invalid area before redrawing. */
-             // gdk_display_sync (gdk_drawable_get_display (window));
-              //g_usleep (70000);
-            //}
-          
 	  save_region = _gdk_windowing_window_queue_antiexpose (window, update_area);
 
 	  if (save_region)
@@ -2918,7 +2900,6 @@
 	      event.expose.count = 0;
 	      event.expose.region = expose_region;
 	      gdk_region_get_clipbox (expose_region, &event.expose.area);
-	      
 	      (*_gdk_event_func) (&event, _gdk_event_data);
 	      
 	      g_object_unref (window);
@@ -2942,6 +2923,7 @@
   gint                     i;
 
 
+  g_assert (region != NULL );
   wimpl = GDK_WINDOW_IMPL_DIRECTFB (paintable);
   impl = (GdkDrawableImplDirectFB *)wimpl;
   impl->buffered = TRUE;
@@ -2987,11 +2969,10 @@
         {
           DFBRegion reg = { impl->paint_region->extents.x1,
                             impl->paint_region->extents.y1,
-                            impl->paint_region->extents.x2 - 1,
-                            impl->paint_region->extents.y2 - 1 };
-
-          _gdk_directfb_update (impl, &reg);
+                            impl->paint_region->extents.x2 ,
+                            impl->paint_region->extents.y2 };
 
+          impl->surface->Flip(impl->surface, &reg,0);
           gdk_region_destroy (impl->paint_region);
           impl->paint_region = NULL;
         }
