Index: gtk+2.0-2.10.12/gdk/directfb/gdkcolor-directfb.c
===================================================================
--- gtk+2.0-2.10.12.orig/gdk/directfb/gdkcolor-directfb.c	2007-05-02 18:27:33.000000000 +0200
+++ gtk+2.0-2.10.12/gdk/directfb/gdkcolor-directfb.c	2007-05-03 15:09:55.000000000 +0200
@@ -38,12 +38,10 @@
 #include <stdlib.h>
 #include <string.h>
 
+#include "gdkcolor.h"
+#include "gdkinternals.h"
 #include "gdkdirectfb.h"
 #include "gdkprivate-directfb.h"
-
-#include "gdkinternals.h"
-
-#include "gdkcolor.h"
 #include "gdkalias.h"
 
 
@@ -618,6 +616,5 @@
 
   private->info[255].ref_count++;
 }
-
 #define __GDK_COLOR_X11_C__
 #include "gdkaliasdef.c"
Index: gtk+2.0-2.10.12/gdk/directfb/gdkwindow-directfb.c
===================================================================
--- gtk+2.0-2.10.12.orig/gdk/directfb/gdkwindow-directfb.c	2007-05-02 18:27:33.000000000 +0200
+++ gtk+2.0-2.10.12/gdk/directfb/gdkwindow-directfb.c	2007-05-03 15:09:55.000000000 +0200
@@ -2993,6 +2993,12 @@
   iface->process_updates = gdk_window_impl_directfb_process_updates;
 }
 
+void
+gdk_window_beep (GdkWindow *window)
+{
+  gdk_display_beep (gdk_display_get_default());
+}
+
 #define __GDK_WINDOW_X11_C__
 #include "gdkaliasdef.c"
 
Index: gtk+2.0-2.10.12/gdk/directfb/Makefile.am
===================================================================
--- gtk+2.0-2.10.12.orig/gdk/directfb/Makefile.am	2007-05-02 18:27:33.000000000 +0200
+++ gtk+2.0-2.10.12/gdk/directfb/Makefile.am	2007-05-03 15:09:55.000000000 +0200
@@ -4,12 +4,11 @@
 
 INCLUDES = 		\
 	-DG_LOG_DOMAIN=\"Gdk-DirectFB\"	\
+	-DGDK_COMPILATION \
 	-I$(top_srcdir)			\
 	-I$(top_srcdir)/gdk		\
 	-I$(top_builddir)/gdk		\
 	-DG_DISABLE_DEPRECATED		\
-	-DGDK_PIXBUF_DISABLE_DEPRECATED	\
-	-DGDK_DISABLE_DEPRECATED	\
 	@GTK_DEBUG_FLAGS@		\
 	@GDK_DEP_CFLAGS@		
 
