Index: gtk+2.0-2.10.5/gtk/Makefile.am
===================================================================
--- gtk+2.0-2.10.5.orig/gtk/Makefile.am	2006-10-02 16:58:28.000000000 +0200
+++ gtk+2.0-2.10.5/gtk/Makefile.am	2006-10-02 16:58:47.000000000 +0200
@@ -1104,7 +1104,7 @@
 	   && $(LN_S) gtk-goto-first-ltr.png gtk-goto-last-rtl.png		\
 	   && $(RM) gtk-media-forward-rtl.png 					\
 	   && $(LN_S) gtk-media-rewind-ltr.png gtk-media-forward-rtl.png	\
-	   && $(RM) gtk-mdedia-next-rtl.png					\
+	   && $(RM) gtk-media-next-rtl.png					\
 	   && $(LN_S) gtk-media-previous-ltr.png gtk-media-next-rtl.png		\
 	   && $(RM) gtk-media-previous-rtl.png 					\
 	   && $(LN_S) gtk-media-next-ltr.png gtk-media-previous-rtl.png		\
