Index: gtk+2.0-2.10.10/gdk/directfb/gdkproperty-directfb.c
===================================================================
--- gtk+2.0-2.10.10.orig/gdk/directfb/gdkproperty-directfb.c	2007-03-12 15:08:51.000000000 +0100
+++ gtk+2.0-2.10.10/gdk/directfb/gdkproperty-directfb.c	2007-03-13 14:10:31.000000000 +0100
@@ -192,7 +192,7 @@
   if (GPOINTER_TO_INT (atom) >= atoms_to_names->len)
     return NULL;
     
-  return g_ptr_array_index (atoms_to_names, GPOINTER_TO_INT (atom));
+  return g_strdup ( g_ptr_array_index (atoms_to_names, GPOINTER_TO_INT (atom)));
 }
 
 
