Index: gtk+2.0-2.11.2/gtk/gtksettings.c
===================================================================
--- gtk+2.0-2.11.2.orig/gtk/gtksettings.c	2007-06-06 13:43:08.000000000 +0200
+++ gtk+2.0-2.11.2/gtk/gtksettings.c	2007-06-13 11:07:39.000000000 +0200
@@ -289,7 +289,7 @@
                                              g_param_spec_string ("gtk-fallback-icon-theme",
 								  P_("Fallback Icon Theme Name"),
 								  P_("Name of a icon theme to fall back to"),
-								  NULL,
+								  "gnome",
 								  GTK_PARAM_READWRITE),
                                              NULL);
   g_assert (result == PROP_FALLBACK_ICON_THEME);
