Index: gtk+-3.2.2/gtk/gtksettings.c
===================================================================
--- gtk+-3.2.2.orig/gtk/gtksettings.c	2011-11-11 23:39:31.000000000 +0100
+++ gtk+-3.2.2/gtk/gtksettings.c	2011-11-13 02:59:24.326906185 +0100
@@ -408,7 +408,7 @@
                                              g_param_spec_string ("gtk-fallback-icon-theme",
                                                                   P_("Fallback Icon Theme Name"),
                                                                   P_("Name of a icon theme to fall back to"),
-                                                                  NULL,
+                                                                  "gnome",
                                                                   GTK_PARAM_READWRITE),
                                              NULL);
   g_assert (result == PROP_FALLBACK_ICON_THEME);
