Index: gtk+-3.20.3/gtk/gtksettings.c
===================================================================
--- gtk+-3.20.3.orig/gtk/gtksettings.c	2016-04-16 20:12:13.342293338 +0200
+++ gtk+-3.20.3/gtk/gtksettings.c	2016-04-16 20:12:13.338293278 +0200
@@ -464,7 +464,7 @@
                                              g_param_spec_string ("gtk-fallback-icon-theme",
                                                                   P_("Fallback Icon Theme Name"),
                                                                   P_("Name of a icon theme to fall back to"),
-                                                                  NULL,
+                                                                  "Adwaita",
                                                                   GTK_PARAM_READWRITE | G_PARAM_DEPRECATED),
                                              NULL);
   g_assert (result == PROP_FALLBACK_ICON_THEME);
