Index: gtk+-3.18.1/gtk/gtksettings.c
===================================================================
--- gtk+-3.18.1.orig/gtk/gtksettings.c	2015-10-08 00:19:24.297948066 +0200
+++ gtk+-3.18.1/gtk/gtksettings.c	2015-10-08 00:19:24.293948010 +0200
@@ -468,7 +468,7 @@
                                              g_param_spec_string ("gtk-fallback-icon-theme",
                                                                   P_("Fallback Icon Theme Name"),
                                                                   P_("Name of a icon theme to fall back to"),
-                                                                  NULL,
+                                                                  "Adwaita",
                                                                   GTK_PARAM_READWRITE | G_PARAM_DEPRECATED),
                                              NULL);
   g_assert (result == PROP_FALLBACK_ICON_THEME);
