Index: b/gtk/gtksettings.c
===================================================================
--- a/gtk/gtksettings.c	2010-04-10 14:52:04.289945999 +0200
+++ b/gtk/gtksettings.c	2010-04-10 14:52:09.473946523 +0200
@@ -310,7 +310,7 @@
                                              g_param_spec_string ("gtk-fallback-icon-theme",
 								  P_("Fallback Icon Theme Name"),
 								  P_("Name of a icon theme to fall back to"),
-								  NULL,
+								  "gnome",
 								  GTK_PARAM_READWRITE),
                                              NULL);
   g_assert (result == PROP_FALLBACK_ICON_THEME);
