Index: b/modules/input/imviqr.c
===================================================================
--- a/modules/input/imviqr.c	2010-04-10 14:52:04.009941613 +0200
+++ b/modules/input/imviqr.c	2010-04-10 14:52:10.389945321 +0200
@@ -244,7 +244,7 @@
   N_("Vietnamese (VIQR)"), /* Human readable name */
   GETTEXT_PACKAGE,	   /* Translation domain */
    GTK_LOCALEDIR,	   /* Dir for bindtextdomain (not strictly needed for "gtk+") */
-  "vi"			   /* Languages for which this module is the default */
+  ""			   /* Languages for which this module is the default */
 };
 
 static const GtkIMContextInfo *info_list[] = {
