Index: gtk+2.0-2.10.6/gtk/gtkentry.c
===================================================================
--- gtk+2.0-2.10.6.orig/gtk/gtkentry.c	2006-12-29 02:08:52.000000000 +0100
+++ gtk+2.0-2.10.6/gtk/gtkentry.c	2006-12-29 02:10:32.000000000 +0100
@@ -521,7 +521,7 @@
                                    g_param_spec_unichar ("invisible-char",
 							 P_("Invisible character"),
 							 P_("The character to use when masking entry contents (in \"password mode\")"),
-							 '*',
+							 (gunichar) 0x25CF,
 							 GTK_PARAM_READWRITE));
 
   g_object_class_install_property (gobject_class,
@@ -1059,7 +1059,7 @@
 
   entry->editable = TRUE;
   entry->visible = TRUE;
-  entry->invisible_char = '*';
+  entry->invisible_char = (gunichar) 0x25CF;
   entry->dnd_position = -1;
   entry->width_chars = -1;
   entry->is_cell_renderer = FALSE;
@@ -4236,9 +4236,9 @@
  * gtk_entry_set_visibility() has been called to set text visibility
  * to %FALSE. i.e. this is the character used in "password mode" to
  * show the user how many characters have been typed. The default
- * invisible char is an asterisk ('*').  If you set the invisible char
- * to 0, then the user will get no feedback at all; there will be
- * no text on the screen as they type.
+ * invisible char is a black circle (Unicode character 25CF).  If
+ * you set the invisible char to 0, then the user will get no
+ * feedback at all; there will be no text on the screen as they type.
  * 
  **/
 void
