Index: gtk+2.0-2.10.4/gtk/gtkfilechooserbutton.c
===================================================================
--- gtk+2.0-2.10.4.orig/gtk/gtkfilechooserbutton.c	2006-09-30 20:03:11.000000000 +0200
+++ gtk+2.0-2.10.4/gtk/gtkfilechooserbutton.c	2006-09-30 20:03:15.000000000 +0200
@@ -2207,10 +2207,13 @@
 	    GtkFilePath *base_path;
 
 	    base_path = gtk_file_system_volume_get_base_path (priv->fs, data);
-	    row_found = (paths &&
-			 paths->data &&
-			 gtk_file_path_compare (base_path, paths->data) == 0);
-	    gtk_file_path_free (base_path);
+            if (base_path)
+              {
+	        row_found = (paths &&
+			     paths->data &&
+			     gtk_file_path_compare (base_path, paths->data) == 0);
+	        gtk_file_path_free (base_path);
+              }
 	  }
 	  break;
 	default:
