Index: gtk+-2.12.0/gtk/gtksearchenginetracker.c
===================================================================
--- gtk+-2.12.0.orig/gtk/gtksearchenginetracker.c	2007-09-21 12:45:21.092692030 +0200
+++ gtk+-2.12.0/gtk/gtksearchenginetracker.c	2007-09-21 12:46:33.856838622 +0200
@@ -35,6 +35,7 @@
 
 static TrackerClient * (*tracker_connect) (gboolean enable_warnings) = NULL;
 static void	       (*tracker_disconnect) (TrackerClient *client) = NULL;
+static int             (*tracker_get_version) (TrackerClient *client, GError **error) = NULL;
 static void            (*tracker_cancel_last_call) (TrackerClient *client) = NULL;
 
 static void (*tracker_search_metadata_by_text_async) (TrackerClient *client, 
@@ -56,6 +57,7 @@
 #define MAP(a) { #a, (gpointer *)&a }
   MAP (tracker_connect),
   MAP (tracker_disconnect),
+  MAP (tracker_get_version),
   MAP (tracker_cancel_last_call),
   MAP (tracker_search_metadata_by_text_async),
   MAP (tracker_search_metadata_by_text_and_location_async),
@@ -74,7 +76,7 @@
       
       done = TRUE;
  
-      tracker = g_module_open ("libtracker.so.0", G_MODULE_BIND_LAZY | G_MODULE_BIND_LOCAL);
+      tracker = g_module_open ("libtrackerclient.so.0", G_MODULE_BIND_LAZY | G_MODULE_BIND_LOCAL);
       if (!tracker)
 	return;
       
@@ -279,6 +281,7 @@
 {
   GtkSearchEngineTracker *engine;
   TrackerClient *tracker_client;
+  GError *err = NULL;
 
   open_libtracker ();
 
@@ -290,6 +293,17 @@
   if (!tracker_client)
     return NULL;
 
+  if (!tracker_get_version)
+    return NULL;
+
+  tracker_get_version (tracker_client, &err);
+
+  if (err != NULL)
+    {
+      g_error_free (err);
+      return NULL;
+    }
+
   engine = g_object_new (GTK_TYPE_SEARCH_ENGINE_TRACKER, NULL);
 
   engine->priv->client = tracker_client;
