--- a/gtk/gtkrc.c
+++ b/gtk/gtkrc.c
@@ -292,6 +292,17 @@
       if (im_module_file)
         result = g_strdup (im_module_file);
       else
+#if defined(__linux__) && ( defined(__i386__) || defined (__x86_64__) || defined(__ia64__) )
+# if defined (__i386__)
+        result = gtk_rc_make_default_dir ("immodules.cache.32");
+# else
+        result = gtk_rc_make_default_dir ("immodules.cache.64");
+# endif
+        /* Prefer compat immodules.cache file if it's usable. */
+        if (g_file_test(result, G_FILE_TEST_EXISTS))
+          return result;
+        g_free (result);
+#endif
         result = gtk_rc_make_default_dir ("immodules.cache");
     }
 
