Index: gtk+-3.2.2/gtk/gtkrc.c
===================================================================
--- gtk+-3.2.2.orig/gtk/gtkrc.c	2011-11-11 23:39:31.000000000 +0100
+++ gtk+-3.2.2/gtk/gtkrc.c	2011-11-13 02:59:28.222906052 +0100
@@ -960,6 +960,17 @@
       if (im_module_file)
         result = g_strdup (im_module_file);
       else
+#if defined(__linux__) && ( defined(__i386__) || defined (__x86_64__) || defined(__ia64__) )
+# if defined (__i386__)
+        result = gtk_rc_make_default_dir ("immodules.cache.32");
+# else
+        result = gtk_rc_make_default_dir ("immodules.cache.64");
+# endif
+        /* Prefer compat immodules.cache file if it's usable. */
+        if (g_file_test(result, G_FILE_TEST_EXISTS))
+          return result;
+        g_free (result);
+#endif
         result = gtk_rc_make_default_dir ("immodules.cache");
     }
 
