Index: gtk+-2.91.7/gtk/gtkrc.c
===================================================================
--- gtk+-2.91.7.orig/gtk/gtkrc.c	2010-12-22 06:33:56.000000000 +1100
+++ gtk+-2.91.7/gtk/gtkrc.c	2010-12-22 13:22:09.653372002 +1100
@@ -293,6 +293,17 @@
       if (im_module_file)
         result = g_strdup (im_module_file);
       else
+#if defined(__linux__) && ( defined(__i386__) || defined (__x86_64__) || defined(__ia64__) )
+# if defined (__i386__)
+        result = gtk_rc_make_default_dir ("immodules.cache.32");
+# else
+        result = gtk_rc_make_default_dir ("immodules.cache.64");
+# endif
+        /* Prefer compat immodules.cache file if it's usable. */
+        if (g_file_test(result, G_FILE_TEST_EXISTS))
+          return result;
+        g_free (result);
+#endif
         result = gtk_rc_make_default_dir ("immodules.cache");
     }
 
