Index: gtk+3.0-2.90.5/gtk/gtktreeview.c
===================================================================
--- gtk+3.0-2.90.5.orig/gtk/gtktreeview.c	2010-07-10 03:00:19.000000000 +0200
+++ gtk+3.0-2.90.5/gtk/gtktreeview.c	2010-07-30 19:12:47.187084280 +0200
@@ -9588,7 +9588,8 @@
 
       if (cursor_path)
 	{
-	  if (tree_view->priv->selection->type == GTK_SELECTION_MULTIPLE)
+	  if (tree_view->priv->selection->type == GTK_SELECTION_MULTIPLE ||
+	      tree_view->priv->selection->type == GTK_SELECTION_SINGLE)
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, FALSE, FALSE);
 	  else
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, TRUE, FALSE);
