Index: gtk+2.0-2.21.5/gtk/gtktreeview.c
===================================================================
--- gtk+2.0-2.21.5.orig/gtk/gtktreeview.c	2010-07-12 18:32:54.000000000 +0200
+++ gtk+2.0-2.21.5/gtk/gtktreeview.c	2010-07-22 17:57:20.833053801 +0200
@@ -9591,7 +9591,8 @@
 
       if (cursor_path)
 	{
-	  if (tree_view->priv->selection->type == GTK_SELECTION_MULTIPLE)
+	  if (tree_view->priv->selection->type == GTK_SELECTION_MULTIPLE ||
+	      tree_view->priv->selection->type == GTK_SELECTION_SINGLE)
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, FALSE, FALSE);
 	  else
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, TRUE, FALSE);
