Index: gtk+-2.91.7/gtk/gtktreeview.c
===================================================================
--- gtk+-2.91.7.orig/gtk/gtktreeview.c	2010-12-22 03:36:24.000000000 +1100
+++ gtk+-2.91.7/gtk/gtktreeview.c	2010-12-22 13:23:08.483372001 +1100
@@ -10067,7 +10067,8 @@
 
       if (cursor_path)
 	{
-	  if (gtk_tree_selection_get_mode (tree_view->priv->selection) == GTK_SELECTION_MULTIPLE)
+	  if (gtk_tree_selection_get_mode (tree_view->priv->selection) == GTK_SELECTION_MULTIPLE ||
+	      gtk_tree_selection_get_mode (tree_view->priv->selection) == GTK_SELECTION_SINGLE)
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, FALSE, FALSE);
 	  else
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, TRUE, FALSE);
