Index: gtk+-2.12.11/gtk/gtktreeview.c
===================================================================
--- gtk+-2.12.11.orig/gtk/gtktreeview.c	2008-07-10 19:00:15.873743659 +0200
+++ gtk+-2.12.11/gtk/gtktreeview.c	2008-07-10 19:00:29.729244218 +0200
@@ -9679,7 +9679,8 @@ gtk_tree_view_focus_to_cursor (GtkTreeVi
 
       if (cursor_path)
 	{
-	  if (tree_view->priv->selection->type == GTK_SELECTION_MULTIPLE)
+	  if (tree_view->priv->selection->type == GTK_SELECTION_MULTIPLE ||
+	      tree_view->priv->selection->type == GTK_SELECTION_SINGLE)
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, FALSE, FALSE);
 	  else
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, TRUE, FALSE);
