Index: gtk+-3.0.0/gtk/gtktreeview.c
===================================================================
--- gtk+-3.0.0.orig/gtk/gtktreeview.c	2011-02-10 23:27:24.000000000 +1100
+++ gtk+-3.0.0/gtk/gtktreeview.c	2011-02-14 11:29:32.687354277 +1100
@@ -10120,7 +10120,8 @@
 
       if (cursor_path)
 	{
-	  if (gtk_tree_selection_get_mode (tree_view->priv->selection) == GTK_SELECTION_MULTIPLE)
+	  if (gtk_tree_selection_get_mode (tree_view->priv->selection) == GTK_SELECTION_MULTIPLE ||
+	      gtk_tree_selection_get_mode (tree_view->priv->selection) == GTK_SELECTION_SINGLE)
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, FALSE, FALSE);
 	  else
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, TRUE, FALSE);
