--- gdk/gdkwindow.c.old	2007-11-27 05:31:06.000000000 +0100
+++ gtk+2.0-2.12.2/gdk/gdkwindow.c	2007-11-27 05:33:15.000000000 +0100
@@ -614,9 +614,12 @@
   g_return_val_if_fail (GDK_IS_WINDOW (window), NULL);
 
   obj = (GdkWindowObject *)window;
-  while (GDK_WINDOW_TYPE (obj) == GDK_WINDOW_CHILD)
+  while (obj && GDK_WINDOW_TYPE (obj) == GDK_WINDOW_CHILD) {
     obj = (GdkWindowObject *)obj->parent;
-  
+  }
+
+  g_return_val_if_fail (obj != NULL, NULL);
+
   return GDK_WINDOW (obj);
 }
 
@@ -2295,9 +2298,11 @@
 {
   GdkWindowObject *toplevel;
 
+  g_return_val_if_fail (window != NULL, FALSE);
+
   toplevel = (GdkWindowObject *)gdk_window_get_toplevel (window);
 
-  return toplevel->update_and_descendants_freeze_count > 0;
+  return toplevel ? toplevel->update_and_descendants_freeze_count > 0 : FALSE;
 }
 
 static void
@@ -2433,6 +2438,9 @@
   
   while (tmp_list)
     {
+      if(!tmp_list->data)
+        continue;
+
       GdkWindowObject *private = (GdkWindowObject *)tmp_list->data;
       
       if (!GDK_WINDOW_DESTROYED (tmp_list->data))
