Description: gtkmenu: fix unnecessary scroll buttons
Origin: upstream, https://git.gnome.org/browse/gtk+/commit/?id=bdaef1f88d44ca26
Last-Update: 2014-11-04

--- a/gtk/gtkmenu.c
+++ b/gtk/gtkmenu.c
@@ -3347,6 +3347,7 @@
     {
       GdkScreen *screen = gtk_widget_get_screen (priv->toplevel);
       GdkRectangle monitor;
+      GtkBorder border;
 
       gdk_screen_get_monitor_workarea (screen, priv->monitor_num, &monitor);
 
@@ -3356,10 +3357,12 @@
       if (priv->position_y + nat_height > monitor.y + monitor.height)
         nat_height = monitor.y + monitor.height - priv->position_y;
 
-      if (priv->position_y < monitor.y)
+      _gtk_window_get_shadow_width (GTK_WINDOW (priv->toplevel), &border);
+
+      if (priv->position_y + border.top < monitor.y)
         {
-          min_height -= monitor.y - priv->position_y;
-          nat_height -= monitor.y - priv->position_y;
+          min_height -= monitor.y - (priv->position_y + border.top);
+          nat_height -= monitor.y - (priv->position_y + border.top);
         }
     }
 
