From aa08b87ab67ec05c5547e7dda2dbefbf6c8fca4e Mon Sep 17 00:00:00 2001
From: Rico Tzschichholz <ricotz@ubuntu.com>
Date: Mon, 22 Sep 2014 15:19:51 +0200
Subject: [PATCH] reftest: Add missing include of reftest-snapshot.h

---
 testsuite/reftests/gtk-reftest.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/testsuite/reftests/gtk-reftest.c b/testsuite/reftests/gtk-reftest.c
index 4f5b6f4..42d914f 100644
--- a/testsuite/reftests/gtk-reftest.c
+++ b/testsuite/reftests/gtk-reftest.c
@@ -22,6 +22,7 @@
 
 #include "reftest-compare.h"
 #include "reftest-module.h"
+#include "reftest-snapshot.h"
 
 #include <string.h>
 #include <glib/gstdio.h>
-- 
2.1.1

