--- gdk/directfb/gdkkeys-directfb.c.old	2010-02-24 14:46:25.924954435 +0100
+++ gdk/directfb/gdkkeys-directfb.c	2010-02-24 14:47:03.199594370 +0100
@@ -1998,6 +1998,20 @@
 	return 0;
 }
 
+void
+gdk_keymap_add_virtual_modifiers (GdkKeymap       *keymap,
+                                  GdkModifierType *state)
+{
+  /* FIXME: For now, we've mimiced the Windows backend. */
+}
+
+gboolean
+gdk_keymap_map_virtual_modifiers (GdkKeymap       *keymap,
+                                  GdkModifierType *state)
+{
+  /* FIXME: For now, we've mimiced the Windows backend. */
+  return TRUE;
+}
 
 #define __GDK_KEYS_DIRECTFB_C__
 #include "gdkaliasdef.c"
--- gdk/gdkaliasdef.c.old	2010-02-24 14:46:30.529155696 +0100
+++ gdk/gdkaliasdef.c	2010-02-24 14:47:36.460953699 +0100
@@ -1845,6 +1845,12 @@
 #undef gdk_keymap_lookup_key 
 extern __typeof (gdk_keymap_lookup_key) gdk_keymap_lookup_key __attribute((alias("IA__gdk_keymap_lookup_key"), visibility("default")));
 
+#undef gdk_keymap_add_virtual_modifiers 
+extern __typeof (gdk_keymap_add_virtual_modifiers) gdk_keymap_add_virtual_modifiers __attribute((alias("IA__gdk_keymap_add_virtual_modifiers"), visibility("default")));
+
+#undef gdk_keymap_map_virtual_modifiers 
+extern __typeof (gdk_keymap_map_virtual_modifiers) gdk_keymap_map_virtual_modifiers __attribute((alias("IA__gdk_keymap_map_virtual_modifiers"), visibility("default")));
+
 #undef gdk_keymap_translate_keyboard_state 
 extern __typeof (gdk_keymap_translate_keyboard_state) gdk_keymap_translate_keyboard_state __attribute((alias("IA__gdk_keymap_translate_keyboard_state"), visibility("default")));
 
