Index: gtk+2.0-2.14.3/modules/printbackends/test/gtkprintbackendtest.c
===================================================================
--- gtk+2.0-2.14.3.orig/modules/printbackends/test/gtkprintbackendtest.c	2008-09-24 21:07:22.000000000 -0300
+++ gtk+2.0-2.14.3/modules/printbackends/test/gtkprintbackendtest.c	2008-09-24 21:07:28.000000000 -0300
@@ -35,12 +35,12 @@
 
 #include <glib/gi18n-lib.h>
 
-#include "gtkprintoperation.h"
+#include <gtk/gtk.h>
 
 #include "gtkprintbackend.h"
 #include "gtkprintbackendtest.h"
 
-#include "gtkprinter.h"
+#include <gtk/gtkunixprint.h>
 #include "gtkprinter-private.h"
 
 typedef struct _GtkPrintBackendTestClass GtkPrintBackendTestClass;
