--- gtk+2.0-2.12.4.orig/gdk/directfb/gdkcursor-directfb.c	2008-01-08 05:20:48.000000000 +0100
+++ gtk+2.0-2.12.4/gdk/directfb/gdkcursor-directfb.c	2008-01-08 17:58:13.000000000 +0100
@@ -46,6 +46,19 @@
 #define u8 __u8
 #endif
 
+
+/* Workaround for DirectFB >= 1.0.0-rc2 not
+ * defining __u32 anymore when DIRECTFB_NO_CRUFT
+ * is set, which is the default on Debian systems.
+ */
+#ifndef __u32
+#define __u32 u32
+#endif
+
+#ifndef __u8
+#define __u8 u8
+#endif
+
 static struct {
   const guchar *bits;
   int width, height, hotx, hoty;
