diff -Nur gtk+2.0-2.16.0.ubuntu/gdk/x11/gdkwindow-x11.c gtk+2.0-2.16.0/gdk/x11/gdkwindow-x11.c
--- gtk+2.0-2.16.0.ubuntu/gdk/x11/gdkwindow-x11.c	2009-04-09 10:44:34.000000000 +0200
+++ gtk+2.0-2.16.0/gdk/x11/gdkwindow-x11.c	2009-04-09 10:51:23.000000000 +0200
@@ -3995,7 +3995,9 @@
                    XA_CARDINAL, 32, PropModeReplace,
                    (guchar *)&timestamp_long, 1);
 
-  if (timestamp_long != GDK_CURRENT_TIME)
+  if (timestamp_long != GDK_CURRENT_TIME &&
+      (display_x11->user_time == GDK_CURRENT_TIME ||
+       XSERVER_TIME_IS_LATER (timestamp_long, display_x11->user_time)))
     display_x11->user_time = timestamp_long;
 
   if (toplevel)
diff -Nur gtk+2.0-2.16.0.ubuntu/gtk/gtkwindow.c gtk+2.0-2.16.0/gtk/gtkwindow.c
--- gtk+2.0-2.16.0.ubuntu/gtk/gtkwindow.c	2009-04-09 10:44:34.000000000 +0200
+++ gtk+2.0-2.16.0/gtk/gtkwindow.c	2009-04-09 10:51:30.000000000 +0200
@@ -395,7 +395,8 @@
     
       /* Skip past the "_TIME" part */
       timestr += 5;
-    
+
+      errno = 0;
       timestamp = strtoul (timestr, &end, 0);
       if (end != timestr && errno == 0)
         retval = timestamp;
@@ -1499,18 +1500,21 @@
   
   g_free (priv->startup_id);
   priv->startup_id = g_strdup (startup_id);
-  
+
   if (GTK_WIDGET_REALIZED (window))
     {
+      guint32 timestamp = extract_time_from_startup_id (priv->startup_id);
+
+#ifdef GDK_WINDOWING_X11
+      if (timestamp != GDK_CURRENT_TIME)
+	gdk_x11_window_set_user_time (GTK_WIDGET (window)->window, timestamp);
+#endif
+
       /* Here we differentiate real and "fake" startup notification IDs,
        * constructed on purpose just to pass interaction timestamp
-       */  
+       */
       if (startup_id_is_fake (priv->startup_id))
-        {
-          guint32 timestamp = extract_time_from_startup_id (priv->startup_id);
-
-          gtk_window_present_with_time (window, timestamp);
-        }
+	gtk_window_present_with_time (window, timestamp);
       else 
         {
           gdk_window_set_startup_id (GTK_WIDGET (window)->window,
