--- gtk+2.0-2.12.6.orig/gdk/directfb/gdkdrawable-directfb.c	2008-01-29 17:34:23.000000000 +0100
+++ gtk+2.0-2.12.6/gdk/directfb/gdkdrawable-directfb.c	2008-01-29 17:34:42.000000000 +0100
@@ -477,7 +477,7 @@
                        gint         angle1,
                        gint         angle2)
 {
-  WARN_UNIMPLEMENTED (G_GNUC_FUNCTION);
+  WARN_UNIMPLEMENTED (G_STRLOC);
 }
 
 static void
@@ -561,7 +561,7 @@
                         const gchar *text,
                         gint         text_length)
 {
-  WARN_UNIMPLEMENTED (G_GNUC_FUNCTION);
+  WARN_UNIMPLEMENTED (G_STRLOC);
 }
 
 static void
@@ -573,7 +573,7 @@
                            const GdkWChar *text,
                            gint            text_length)
 {
-  WARN_UNIMPLEMENTED (G_GNUC_FUNCTION);
+  WARN_UNIMPLEMENTED (G_STRLOC);
 }
 
 static void
--- gtk+2.0-2.12.6.orig/gdk/directfb/gdkimage-directfb.c	2008-01-29 17:34:59.000000000 +0100
+++ gtk+2.0-2.12.6/gdk/directfb/gdkimage-directfb.c	2008-01-29 17:35:10.000000000 +0100
@@ -151,7 +151,7 @@
 
   GDK_NOTE (MISC, g_print ("gdk_image_new_bitmap: %dx%d\n", w, h));
 
-  g_message ("not fully implemented %s", G_GNUC_FUNCTION);
+  g_message ("not fully implemented %s", G_STRLOC);
 
   image->bpl = (w + 7) / 8;
   image->mem = g_malloc (image->bpl * h);
@@ -209,7 +209,7 @@
       format = DSPF_ARGB;
       break;
     default:
-      g_message ("unimplemented %s for depth %d", G_GNUC_FUNCTION, depth);
+      g_message ("unimplemented %s for depth %d", G_STRLOC, depth);
       return NULL;
     }
 
--- gtk+2.0-2.12.6.orig/gdk/directfb/gdkinput-directfb.c	2008-01-29 17:35:30.000000000 +0100
+++ gtk+2.0-2.12.6/gdk/directfb/gdkinput-directfb.c	2008-01-29 17:35:38.000000000 +0100
@@ -228,7 +228,7 @@
 gdk_device_set_mode (GdkDevice    *device,
                      GdkInputMode  mode)
 {
-  g_message ("unimplemented %s", G_GNUC_FUNCTION);
+  g_message ("unimplemented %s", G_STRLOC);
 
   return FALSE;
 }
@@ -303,7 +303,7 @@
                                 gint              mask,
                                 GdkExtensionMode  mode)
 {
-  g_message ("unimplemented %s", G_GNUC_FUNCTION);
+  g_message ("unimplemented %s", G_STRLOC);
 }
 
 GList *
--- gtk+2.0-2.12.6.orig/gdk/directfb/gdkpixmap-directfb.c	2008-01-29 17:35:52.000000000 +0100
+++ gtk+2.0-2.12.6/gdk/directfb/gdkpixmap-directfb.c	2008-01-29 17:36:10.000000000 +0100
@@ -172,7 +172,7 @@
           format = DSPF_RGB32;
           break;
         default:
-          g_message ("unimplemented %s for depth %d", G_GNUC_FUNCTION, depth);
+          g_message ("unimplemented %s for depth %d", G_STRLOC, depth);
           return NULL;
         }
     }
--- gtk+2.0-2.12.6.orig/gdk/directfb/gdkwindow-directfb.c	2008-01-29 17:36:26.000000000 +0100
+++ gtk+2.0-2.12.6/gdk/directfb/gdkwindow-directfb.c	2008-01-29 17:36:36.000000000 +0100
@@ -2411,7 +2411,7 @@
     return;
 
   /* N/A */
-  g_message("unimplemented %s", G_GNUC_FUNCTION);
+  g_message("unimplemented %s", G_STRLOC);
 }
 
 void
@@ -2444,7 +2444,7 @@
     return FALSE;
 
   /* N/A */
-  g_message("unimplemented %s", G_GNUC_FUNCTION);
+  g_message("unimplemented %s", G_STRLOC);
 
   return FALSE;
 }
@@ -2462,7 +2462,7 @@
   if (GDK_WINDOW_DESTROYED (window))
     return;
 
-  g_message("unimplemented %s", G_GNUC_FUNCTION);
+  g_message("unimplemented %s", G_STRLOC);
 }
 
 void
@@ -2477,7 +2477,7 @@
   if (GDK_WINDOW_DESTROYED (window))
     return;
 
-  g_message("unimplemented %s", G_GNUC_FUNCTION);
+  g_message("unimplemented %s", G_STRLOC);
 }
 
 /**
