From: Simon McVittie <smcv@debian.org>
Date: Mon, 30 Aug 2021 10:48:49 +0100
Subject: compose: Document how to get compose-parse input from libX11 source

Signed-off-by: Simon McVittie <smcv@debian.org>
Forwarded: https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/3902
Applied-upstream: 4.5.0
---
 gtk/compose/compose-parse.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/gtk/compose/compose-parse.c b/gtk/compose/compose-parse.c
index a1ec09d..385a34f 100644
--- a/gtk/compose/compose-parse.c
+++ b/gtk/compose/compose-parse.c
@@ -7,6 +7,7 @@
  * character data, and definitions for the builtin compose table of GTK.
  * Run it like this:
  *
+ *   cpp -DXCOMM='#' Compose.pre | sed -e 's/^ *#/#/' > Compose
  *   compose-parse Compose sequences-little-endian sequences-big-endian chars gtkcomposedata.h
  *
  * The GTK build expects the output files to be in the source tree, in
